pandas/core/internals/blocks.py

```diff
@@ -863,14 +867,21 @@ def setitem(self, indexer, value):
             #  be e.g. a list; see GH#6043
             values[indexer] = value
 
-        # if we are an exact match (ex-broadcasting),
-        # then use the resultant dtype
         elif (
-            len(arr_value.shape)
-            and arr_value.shape[0] == values.shape[0]
-            and arr_value.size == values.size
+            exact_match
```

Comment:
so i think its possible to greatly simplify this entire routine, but creating a Categorical.set_item. It would do this kind of test and if so just sets (like you have here), otherwise it can dispatch to the Block.set_item (this routine) for regular handling.

The logic in this routine (Block.set_item) then becomes a lot simpler.

now it might be slightly non-trivial to change this because .set_item does a lot, e.g. it preps the values AND then does the setting; these likely need to be decoupled to avoid duplicating some code.

lmk if this is doable and you want to work on it. ok with merging this if we can followup with the Categorical.set_item change (or just do that),.

Reply:
Hi @jreback, sorry for the delay. I would prefer to merge this, and then follow up with the Categorical.set_item change. 

Merged master, and updated.

Reply:
k let's do that